src/test/java/com/impactupgrade/nucleus/it/MailchimpIT.java

```diff
@@ -0,0 +1,125 @@
+    List<MemberInfo> listMembers = mailchimpClient.getListMembers(listId, "subscribed", "members.email_address,members.status,total_items", null);
+    assertNotNull(listMembers);
+
+    List<String> listMembersEmails = listMembers.stream().map(m -> m.email_address).collect(Collectors.toList());
+    assertTrue(listMembersEmails.containsAll(unsubscribeEmails));
+    for (MemberInfo memberInfo : listMembers) {
+      if (emails.contains(memberInfo.email_address)) {
+        assertEquals("subscribed", memberInfo.status);
+      }
+    }
+
+    // 3. Set "unsubscribe" status
+    updateEmailsStatus(unsubscribeEmails, "unsubscribed", listId, mailchimpClient);
+    listMembers = mailchimpClient.getListMembers(listId, "unsubscribed", "members.email_address,members.status,total_items", null);
+    for (MemberInfo memberInfo : listMembers) {
+      if (unsubscribeEmails.contains(memberInfo.email_address)) {
+        assertEquals("unsubscribed", memberInfo.status);
+      }
+    }
+
+    // 3. Set "cleaned" status
+    updateEmailsStatus(cleanEmails, "cleaned", listId, mailchimpClient);
+    listMembers = mailchimpClient.getListMembers(listId, "cleaned", "members.email_address,members.status,total_items", null);
+    for (MemberInfo memberInfo : listMembers) {
+      if (cleanEmails.contains(memberInfo.email_address)) {
+        assertEquals("cleaned", memberInfo.status);
+      }
+    }
+  }
+
+  // Utils
+  private List<String> toCsvList(String email) {
+    // firstname, lastname, email
+    return List.of(RandomStringUtils.randomAlphabetic(8), RandomStringUtils.randomAlphabetic(8), email);
+  }
+
+  private void addEmailsToList(List<String> emails, String status, String listId, MailchimpClient mailchimpClient) throws Exception {
+    List<MemberInfo> memberInfos = emails.stream().map(email -> toMemberInfo(email, status)).collect(Collectors.toList());
+    mailchimpClient.upsertContactsBatch(listId, memberInfos);
+  }
+
+  private void updateEmailsStatus(List<String> emails, String status, String listId, MailchimpClient mailchimpClient) throws Exception {
+    List<MemberInfo> memberInfos = mailchimpClient.getListMembers(listId, null, "members.email_address,members.status,members.merge_fields,members.interests,total_items", null);
+    Map<String, MemberInfo> memberInfoMap = memberInfos.stream()
+            .collect(Collectors.toMap(
+                    m -> m.email_address, m -> m
+            ));
+    // TODO: replace with bulk update?
+    for (String email : emails) {
+      MemberInfo memberInfo = memberInfoMap.get(email);
+      if (memberInfo == null) {
+        throw new IllegalStateException("Failed to find email '" + email + "' in list '" + listId + "'! Skipping update.");
+      } else {
+        memberInfo.status = status;
+        memberInfo.interests = new MailchimpObject(); // API does not return this field for some reason but is required for update
+        mailchimpClient.upsertContact(listId, memberInfo);
```

Comment:
Same comment as above. The goal of this IT is to ensure the CRM<->MC sync is working through MailchimpCommunicationService. Instead, the test bulk inserts contacts to the CRM (further up), but then appears to be doing nothing else with them? Everything else is purely upserting contacts directly through the MC API, skipping the sync entirely. Right?

Reply:
Correct!
Will implement the mentioned flow:
rather than using the MC API directly to insert the contacts, first add them to the CRM with their opt-ins checked. Then run the MC sync and ensure those contacts get subscribed to the MC list. After those assertions, then update the contacts to cleaned/unsubscribed, resync, and ensure the appropriate action is taken